apache/incubator-reef
# [REEF-30] Run REEF on Mesos

JIRA: [REEF-30]: https://issues.apache.org/jira/browse/REEF-30


## reef_on_mesos

#### What's in it
- reef-runtime-mesos package
- bin/runmesostests.sh (for running reef-tests on Mesos clusters)
- HelloREEFMesos in reef-examples
- MesosTestEnvironment in reef-tests
- pom.xml changes(adding reef-runtime-mesos) in root, reef-tests, reef-examples

#### How to test(reef-tests) it

Pretty similar to running reef-tests on YARN.

1. run mesos daemons (http://mesos.apache.org/gettingstarted/)
2. At $REEF_HOME, "mvn clean install -DskipTests"
3. At $REEF_HOME/reef-tests, "mvn jar:test-jar"
4. At $REEF_HOME, "bin/runmesostests.sh $YOUR_MESOS_MASTER_IP" 

Note that you may need to do additional tasks in installing Mesos depending on your platform. For example, on my Mac, I have "export MESOS_NATIVE_LIBRARY=/usr/local/lib/libmesos-0.21.0.dylib" and configured Mesos with "../configure --with-svn=/usr/local/Cellar/subversion/1.8.10_1" command. 

Google is your friend :)


#### reef-runtime-mesos package structure
Client

* MesosClientConfiguration - A ConfigurationModule for the Mesos resource manager
* MesosJobSubmissionHandler - A JobSubmissionHandler that configures and launches the driver to run on Mesos.

Driver

* MesosDriverConfiguration - Binds Driver's runtime event handlers
* MesosOfferManager - Mesos Scheduler that communicates with MesosMaster and MesosExecutor for REEF Driver
* MesosResource(Request/Release/Launch)Handler - Handles requests from REEF driver.
* MesosRuntime(Start/Stop)Handler - Handler of RuntimeStart/RuntimeStop for the MESOS Runtime.

Evaluator

* EvaluatorLaunchHandler - Handles evaluator launch requests from the driver
* EvaluatorReleaseHandler - Handles evaluator release requests from the driver
* MesosExecutor - Mesos Executor that oversees the lifecycle of a REEF Evaluator process


#### Open issues

As this package is still in development stage, many open issues remain. They will be registered as JIRA issues once this PR is merged.

- Launch REEF Driver on a slave node in the cluster
- onRuntimeStop Mesos driver stops before receiving TASK_FINISHED (confusing status on web UI)
- For NodeDescriptorProtos, Mesos Slave's ports are statically defined (5051)
- Use links, not copies in setting up Driver folder structure (there was a similar issue for reef-runtime-local)
- Create tests to cover corner cases not covered by reef-tests
- Support HA REEF Driver
- Support rack/node locality
- Support prioritized requests
- Propose making reef-runtime-local/process package public such that it can be also used for spawning Mesos Driver/Mesos Evaluator processes
- Integrate reef-tests on Mesos into CI
- Explore better ways of mapping Mesos Executor/Task to REEF Evaluator/Task
- Replace the current ad-hoc MesosRemoteManager with Tang's namespace magic
- Research on reserving unused offers instead of declining them
